### PR TITLE
Support libdxcompiler.so on Linux

### PR DESCRIPTION
**Pull request by MarijnS95**
Thanks @h3r2tic for doing the initial porting work!

Things to sort out in the feature:
- Make vtables binary-compatible with Windows again (requires changes in the upstream project);
- Update deprecated `com-rs` crate to Microsofts' com-rs. This is in the works but depends on [minor changes in com-rs](https://github.com/microsoft/com-rs/pull/153) to compile on Linux. That project is furthermore going through an API refactor which makes it hard to justify switching just yet.

### CODE REVIEW

File: examples/validate.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ pub struct MinimalHeader {
 
 // zero_digest & get_digest from https://github.com/gwihlidal/dxil-signing/blob/master/rust/src/main.rs
 
-fn zero_digest(buffer: &mut [u8]) -> () {
+fn zero_digest(buffer: &mut [u8]) {
     let buffer_ptr: *mut u8 = buffer.as_mut_ptr();
     let header_ptr: *mut MinimalHeader = buffer_ptr as *mut _;
     let header_mut: &mut MinimalHeader = unsafe { &mut *header_ptr };
```
Comment (MarijnS95) [inline examples/validate.rs:14]: Clippy complains about an alignment mismatch on `*u8` vs `*MinimalHeader` (`1 < 4`). Should we change the repr of the latter to `repr(packed)`?

File: src/ffi.rs [hunk 2]
```diff
@@ -1,34 +1,34 @@
...
 use std::ffi::c_void;
-use winapi::shared::ntdef::{LPCWSTR, LPWSTR};
-use winapi::shared::winerror::HRESULT;
 
 pub type DxcCreateInstanceProc =
     extern "system" fn(rclsid: &IID, riid: &IID, ppv: *mut *mut c_void) -> HRESULT;
 
 pub type DxcCreateInstanceProc2 = extern "system" fn(
-    malloc: *const c_void,
+    malloc: /* IMalloc */ *const c_void,
```
Comment (MarijnS95) [inline src/ffi.rs:13]: Perhaps we should make these `pub type IMalloc = c_void;` (and `pub type IStream = c_void;` below) instead?

Comment (MarijnS95) [reply src/ffi.rs:13]: They likely won't be used any time soon, but it's good to have all non-void types (in the actual DXC API) properly represented even if the backing object isn't implemented.

File: src/wrapper.rs [hunk 3]
```diff
@@ -425,11 +434,23 @@ impl DxcLibrary {
...
+                self.inner.create_blob_with_encoding_on_heap_copy(
+                    text.as_ptr() as *const c_void,
+                    text.len() as u32,
+                    CP_UTF8,
+                    blob.as_mut_ptr(),
+                )
+            },
+            DxcBlobEncoding::new(blob)
+        );*/
+
```
Comment (MarijnS95) [inline src/wrapper.rs:453]: Looks like this should disappear.

File: examples/validate.rs (hunk 1 above)
Comment (Jasper-Bekkers) [reply examples/validate.rs:15]: Done

Comment (MarijnS95) [reply examples/validate.rs:15]: The `#[repr(C)]` can/should disappear then, right?

File: src/utils.rs [hunk 4]
```diff
@@ -69,6 +69,10 @@ pub enum HassleError {
     CompileError(String),
     #[error("Validation error: {0}")]
     ValidationError(String),
+    #[error("Error loading library")]
+    LoadLibraryError(#[from] libloading::Error),
```
Comment (MarijnS95) [inline src/utils.rs]: Is the underlying error printed in this case, or is that only applicable to `#[error(transparent)]`?

Comment (Jasper-Bekkers) [reply src/utils.rs]: Updated, because this also wouldn't print the filename of the DLL / SO that was being attempted.

**Pull request merged**